lib/auth/password_test.go

```diff
@@ -226,76 +226,167 @@ func (s *PasswordSuite) TestChangePasswordWithOTP(c *C) {
+				c.Assert(err, IsNil)
+			},
+			getReq: func(resetTokenID string) *proto.ChangeUserAuthenticationRequest {
+				u2fRegResp, _, err := getMockedU2FAndRegisterRes(s.a, resetTokenID)
+				c.Assert(err, IsNil)
+
+				return &proto.ChangeUserAuthenticationRequest{
+					TokenID:     resetTokenID,
+					NewPassword: []byte("password3"),
+					NewMFARegisterResponse: &proto.MFARegisterResponse{Response: &proto.MFARegisterResponse_U2F{
+						U2F: u2fRegResp,
+					}},
+				}
+			},
+			// Invalid totp fields when auth settings set to only u2f.
+			getInvalidReq: func(resetTokenID string) *proto.ChangeUserAuthenticationRequest {
+				return &proto.ChangeUserAuthenticationRequest{
+					TokenID:                resetTokenID,
+					NewPassword:            []byte("password3"),
+					NewMFARegisterResponse: &proto.MFARegisterResponse{Response: &proto.MFARegisterResponse_TOTP{}},
+				}
+			},
+		},
+		{
+			name: "with second factor on",
+			setAuthPreference: func() {
+				authPreference.SetSecondFactor(constants.SecondFactorOn)
+				err = s.a.SetAuthPreference(ctx, authPreference)
+				c.Assert(err, IsNil)
+			},
+			getReq: func(resetTokenID string) *proto.ChangeUserAuthenticationRequest {
+				u2fRegResp, _, err := getMockedU2FAndRegisterRes(s.a, resetTokenID)
+				c.Assert(err, IsNil)
+
+				return &proto.ChangeUserAuthenticationRequest{
+					TokenID:     resetTokenID,
+					NewPassword: []byte("password4"),
+					NewMFARegisterResponse: &proto.MFARegisterResponse{Response: &proto.MFARegisterResponse_U2F{
+						U2F: u2fRegResp,
+					}},
+				}
+			},
+			// Empty register response, when auth settings requires second factors.
+			getInvalidReq: func(resetTokenID string) *proto.ChangeUserAuthenticationRequest {
+				return &proto.ChangeUserAuthenticationRequest{
+					TokenID:     resetTokenID,
+					NewPassword: []byte("password4"),
+				}
+			},
+		},
+		{
+			name: "with second factor optional and no second factor",
+			setAuthPreference: func() {
+				authPreference.SetSecondFactor(constants.SecondFactorOptional)
+				err = s.a.SetAuthPreference(ctx, authPreference)
+				c.Assert(err, IsNil)
+			},
+			getReq: func(resetTokenID string) *proto.ChangeUserAuthenticationRequest {
+				return &proto.ChangeUserAuthenticationRequest{
+					TokenID:     resetTokenID,
+					NewPassword: []byte("password5"),
+				}
+			},
+		},
+	}
 
-	username := "joe@example.com"
-	password := []byte("qweqweqwe")
-	_, _, err = CreateUserAndRole(s.a, username, []string{username})
-	c.Assert(err, IsNil)
+	for _, t := range tests {
+		comment := check.Commentf("test case %q", t.name)
 
-	token, err := s.a.CreateResetPasswordToken(ctx, CreateUserTokenRequest{
-		Name: username,
-	})
-	c.Assert(err, IsNil)
+		t.setAuthPreference()
 
-	secrets, err := s.a.RotateUserTokenSecrets(ctx, token.GetName())
-	c.Assert(err, IsNil)
+		token, err := s.a.CreateResetPasswordToken(context.TODO(), CreateUserTokenRequest{
+			Name: username,
```

Comment:
Create distinct users for each test, so it's easier to reason that they don't have previously-registered devices?

Reply:
The `CreateResetPasswordToken` request will reset/remove the users MFA devices

Reply:
actually, i created new user for each test so i can also parallelize it